core/src/advice/mod.rs

```diff
@@ -1 +1,169 @@
-pub(super) mod map;
+use alloc::vec::Vec;
+
+use crate::{
+    crypto::{hash::RpoDigest, merkle::MerklePath},
+    Felt, Word,
+};
+
+mod errors;
+pub use errors::AdviceProviderError;
+
+mod inputs;
+pub use inputs::AdviceInputs;
+
+pub(crate) mod map;
+
+mod source;
+pub use source::AdviceSource;
+
+// ADVICE PROVIDER
+// ================================================================================================
+
+/// Defines behavior of an advice provider.
+///
+/// An advice provider is a component through which the host can interact with the advice provider.
+/// The host can request nondeterministic inputs from the advice provider (i.e., result of a
+/// computation performed outside of the VM), as well as insert new data into the advice provider.
+///
+/// An advice provider consists of the following components:
+/// 1. Advice stack, which is a LIFO data structure. The processor can move the elements from the
+///    advice stack onto the operand stack, as well as push new elements onto the advice stack.
+/// 2. Advice map, which is a key-value map where keys are words (4 field elements) and values are
+///    vectors of field elements. The processor can push the values from the map onto the advice
+///    stack, as well as insert new values into the map.
+/// 3. Merkle store, which contains structured data reducible to Merkle paths. The VM can request
+///    Merkle paths from the store, as well as mutate it by updating or merging nodes contained in
+///    the store.
+pub trait AdviceProvider: Sized {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
```

Comment:
nit: I'd probably get rid of this section separator as all the methods below are required.

Reply:
Removed